### DISCUSSION TITLE
How to create a nested layout?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#8207] [Q&A]
I am trying to use standard razor view within OrchardCms.

I want to create a sub-layout inside the `/Areas/TheAdmin/Views/Layout.cshtml`

My sub-layout needs to look like this

```
<div class="card text-center">
    <div class="card-header">

        <div class="card mb-3">
            <!-- Render client summary here -->
            <!-- This section will show on all the page that use this layout -->
        </div>

        <!-- Render profile navbar here -->
        <!-- This section will show on all the page that use this layout -->
    </div>
    <div class="card-body">
        <!-- Render views here -->
    </div>
</div>
```

I create a layout called "ProfileLayout.cshtml" which contains the following

```
@{
    Layout = "/Areas/TheAdmin/Views/Layout.cshtml";
}
<div class="card text-center">
    <div class="card-header">

        <div class="card mb-3">
            <!-- Render client summary here -->
            <!-- This section will show on all the page that use this layout -->
        </div>

        <!-- Render profile navbar here -->
        <!-- This section will show on all the page that use this layout -->
    </div>
    <div class="card-body">
        @await RenderBodyAsync()
    </div>
</div>
```

Then I created a view called `~/Views/Clients/Index.cshtml` which looks like this

```
@{
     Layout = "ProfileLayout";
}

Some content...
```

When I run the code I get the this error

```
InvalidOperationException: RenderBody has not been called for the page at '/Areas/TheAdmin/Views/Layout.cshtml'. To ignore call IgnoreBody().
```

How can I correct this issue and create layout inside of the main admin layout?

**Comment by hishamco** [comment]
Your layout should have `@RenderBody()`

**Reply by MikeAlhayek** [reply]
That would just raise the same error in the `/Areas/TheAdmin/Views/Layout.cshtml`

**Comment by MikeAlhayek** [comment]
After more research, I found #5576. I created a view called `~/Views/Layout-ClientProfile.cshtml` with the following content 

```
<div class="card text-center">
    <div class="card-header">

        <div class="card mb-3">
            <!-- Render client summary here -->
            <!-- This section will show on all the page that use this layout -->
        </div>

        <!-- Render profile navbar here -->
        <!-- This section will show on all the page that use this layout -->
    </div>
    <div class="card-body">
        @await RenderBodyAsync()
    </div>
</div>
```

Then I changed my view ~/Views/Clients/Index.cshtml to this

```
@{
    ViewLayout = "Layout__ClientProfile";
}

some content from the sub-view
```

This now renders the view content inside the `Layout-ClientProfile.cshtml` "layout". But the `Layout-ClientProfile.cshtml` is not being rendered inside the main `/Areas/TheAdmin/Views/Layout.cshtml` layout.

Adding the following to the `Layout-ClientProfile.cshtml` throws my original error 

```@{ 
    Layout = "/Areas/TheAdmin/Views/Layout.cshtml";
}
```

>> InvalidOperationException: RenderBody has not been called for the page at '/Areas/TheAdmin/Views/Layout.cshtml'. To ignore call IgnoreBody().


Also adding the following to the `Layout-ClientProfile.cshtml` does not render the main layout

```@{ 
    ViewLayout = "/Areas/TheAdmin/Views/Layout.cshtml";
}
```

**Reply by MikeAlhayek** [reply]
@hishamco check my latest progress

**Reply by hishamco** [reply]
@CrestApps do you want to replace the main layout or just adding a nested layout?

**Reply by MikeAlhayek** [reply]
@hishamco  
I don't want to replace the main one. I want to created a nested layout. In other words, I want to render `~/Views/Clients/Index.cshtml` view inside `Layout-ClientProfile.cshtml` layout. Then render `Layout-ClientProfile.cshtml` layout inside of "/Areas/TheAdmin/Views/Layout.cshtml" layout.

**Reply by hishamco** [reply]
Have a look to `ViewLayout` like what Login page is have https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Users/Views/Account/Login.cshtml

**Reply by MikeAlhayek** [reply]
@hishamco  The code [ViewLayout](https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Users/Views/Account/Login.cshtml#L11) is looking for [Layout-Login.cshtml](https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Themes/TheAdmin/Views/Layout-Login.cshtml) layout. Which is exactly what I am doing. 

However, in [Layout-Login.cshtml](https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Themes/TheAdmin/Views/Layout-Login.cshtml) layout is a complete layout not a nested layout. The missing part for me is how would I render [Layout-Login.cshtml](https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Themes/TheAdmin/Views/Layout-Login.cshtml) layout inside [Layout.cshtml](https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Themes/TheAdmin/Views/Layout.cshtml)

**Reply by hishamco** [reply]
Ya `Layout-Login.cshtml` is a complete layout to make it nested just use is it as normal view and set the `Layout` property with `RenderBody` included if it doesn't work with you, I will try it in my source

**Reply by MikeAlhayek** [reply]
@hishamco that does not work. I get the original error 
>> InvalidOperationException: RenderBody has not been called for the page at '/Areas/TheAdmin/Views/Layout.cshtml'. To ignore call IgnoreBody().

**Reply by hishamco** [reply]
I need to test it ..

**Reply by MikeAlhayek** [reply]
@hishamco I think instead of `Layout-ClientProfile` I need to create a wrapper. How can I add a wrapper directly from `~/Views/Clients/Index.cshtml`? `Model` is null I was thinking if that was a shape I would be able to add a wrapper to it

**Reply by hishamco** [reply]
Nested layouts is something for Razor not OC, just few minutes coz my laptop is hanging with OC ;)

**Reply by hishamco** [reply]
Seems nested layouts not supported through ViewLayout

**Reply by hishamco** [reply]
@CrestApps there are differences beween Razor layout & OC layout - which is a view has custom `ViewLayout` - so let us wait for the guys who worked on this feature

**Reply by Skrypt** [reply]
We don't use Nested Layouts mainly because we have Shapes ... but this is somehow an issue that could maybe be fixed.

**Reply by MikeAlhayek** [reply]
@Skrypt I with this issue can be fixed. I’ll save me so much trouble

**Reply by jtkech** [reply]
@CrestApps Quite old but maybe helpfull #993

**Reply by hishamco** [reply]
> but this is somehow an issue that could maybe be fixed.

This will be a good enhancement especially for those how like Razor layouts & views and have less knowledge about shapes

**Reply by ns8482e** [reply]
@CrestApps this is how I am creating nested sub layouts.  See of that fits your needs

https://github.com/OrchardCMS/OrchardCore/issues/3963#issuecomment-580754019
